src/radical/pilot/raptor/worker_mpi.py

```diff
@@ -547,25 +549,43 @@ def _dispatch_function(self, task, env):
               unnamed argument.
         '''
 
-        uid       = task['uid']
-        func_name = task['description']['function']
-        assert(func_name)
+        uid  = task['uid']
+        func = task['description']['function']
 
-        # check if `func_name` is a global name
-        names   = dict(list(globals().items()) + list(locals().items()))
-        to_call = names.get(func_name)
+        to_call = ''
+        args    = task['description'].get('args',   [])
+        kwargs  = task['description'].get('kwargs', {})
+
+        # check if we have a serialized object
+        try:
+            func_info  = deserialize_bson(func)
+            to_call    = deserialize_obj(func_info['func'])
+            args       = func_info["args"]
+            kwargs     = func_info["kwargs"]
```

Comment:
why do we need to have two functions here, could we have one?
```python
func_info = deserialize_func(func)
# func_info could be either tuple `(to_call, args, kwargs)` or dict
```
and I would think the same when we do serialization
```python
return serialize_func(f, args, kwargs)
```

Reply:
@mtitov  They both do `serialize/deserialize` but one returns `string` (`deserialize_bson` which MongoDB accepts)  and one returns `byte` so `deserialize_obj` can be used to deserialize different types of objects which might be needed for use cases like `Colmena`. Combining both is not possible at the moment.

Reply:
So, I figured out what I didn't like (and my comment below could be skipped) - class `PythonTask` (aka FuncSer) is responsible for function serialization, THUS it should be responsible for deserialization (since object was packed here, this class **knows** its structure)

A corresponding method could be static
```python
class PythonTask:
...
    @staticmethod
    def deserialize(f):
        func_info = deserialize_bson(f)
        to_call   = deserialize_obj(func_info['func'])
        return to_call, func_info['args'], func_info['kwargs']
```
and use it as
```python
            to_call, args, kwargs = PythonTask.deserialize(func)
```

Reply:
@mtitov  Why do we want to expose the mechanism of wrapping the python functions to the external API? The purpose of this class `Python Task` is to act as a wrapper, not a wrapper/deserializer. Serializing an object might fall in the category of wrapping but the opposite is not valid IMHOP.

Reply:
> Why do we want to expose the mechanism of wrapping the python functions to the external API?

It is exposed now when you deserialize func-object outside of `PythonTask`. Kinda when you received an object, which you know was serialized by `PythonTask`, you shouldn't deserialize it by yourself.

A simple example: you have several places where you do deserialization of func-object and serialization process got extended with an extra `dumps` call, then you need to fix deserialization at every place it is done. Thus you have to carry knowledge about [de]serialization process inside `PythonTask` and outside, but it should be only within `PythonTask`.

> The purpose of this class `PythonTask` is to act as a wrapper, not a wrapper/deserializer.

`PythonTask` does wrapping of func and produces a serialized func-object, thus it knows the order of `loads` calls and a structure of a serialized object (`{'func': .., 'args': .., 'kwargs': ..}`). You shouldn't move this knowledge outside of the class.

Reply:
Fixed in the latest commit.